### PR TITLE
added primitive logging support

### PR DESCRIPTION
**Pull request by aanurraj**
## Affected Dependencies
List any dependencies that are required for this change.

## How has this been tested?
- Describe the tests that you ran to verify your changes.
- Provide instructions so we can reproduce.
- List any relevant details for your test configuration.

## Checklist
- [ ] I have followed the [Contribution Guidelines](https://github.com/OpenMined/.github/blob/master/CONTRIBUTING.md) and [Code of Conduct](https://github.com/OpenMined/.github/blob/master/CODE_OF_CONDUCT.md)
- [ ] I have commented my code following the [OpenMined Styleguide](https://github.com/OpenMined/.github/blob/master/STYLEGUIDE.md)
- [ ] I have labeled this PR with the relevant [Type labels](https://github.com/OpenMined/.github/labels?q=Type%3A)
- [ ] My changes are covered by tests

### CODE REVIEW

File: src/sympc/store/crypto_primitive_provider.py [hunk 1]
```diff
@@ -13,6 +14,8 @@ class CryptoPrimitiveProvider:
     """A trusted third party should use this class to generate crypto primitives."""
 
     _func_providers: Dict[str, Callable] = {}
+    _logging = False
```
Comment (gmuraru) [inline src/sympc/store/crypto_primitive_provider.py:17]: NITPICK: Add types for this and the ```_ops_list```

File: src/sympc/store/crypto_primitive_provider.py [hunk 2]
```diff
@@ -43,9 +46,18 @@ def generate_primitives(
             raise ValueError(f"{op_str} not registered")
 
         generator = CryptoPrimitiveProvider._func_providers[op_str]
-
         primitives = generator(**g_kwargs)
 
+        if CryptoPrimitiveProvider._logging and (
+            p_kwargs.get("a_shape") and p_kwargs.get("b_shape")
```
Comment (gmuraru) [inline src/sympc/store/crypto_primitive_provider.py:52]: Q: Can't we always presume we have ```a_shape``` and ```b_shape```?

File: src/sympc/store/crypto_primitive_provider.py [hunk 3]
```diff
@@ -13,6 +14,8 @@ class CryptoPrimitiveProvider:
     """A trusted third party should use this class to generate crypto primitives."""
 
     _func_providers: Dict[str, Callable] = {}
+    _logging = False
+    _ops_list = {}
```
Comment (gmuraru) [inline src/sympc/store/crypto_primitive_provider.py:18]: MAYBE: Make it a ```defaultdict``` and on line 54 you won't need to check if it has the element

File: src/sympc/store/crypto_primitive_provider.py (hunk 1 above)
Comment (aanurraj) [reply src/sympc/store/crypto_primitive_provider.py:18]: done

File: src/sympc/store/crypto_primitive_provider.py (hunk 3 above)
Comment (aanurraj) [reply src/sympc/store/crypto_primitive_provider.py]: done

File: src/sympc/store/crypto_primitive_provider.py (hunk 2 above)
Comment (aanurraj) [reply src/sympc/store/crypto_primitive_provider.py:53]: I wasnt sure about it... do we have any other way/idea ?

Comment (gmuraru) [reply src/sympc/store/crypto_primitive_provider.py:53]: If we simply remove them, do they fail?

Comment (gmuraru) [reply src/sympc/store/crypto_primitive_provider.py:53]: Discussed on slack :D

File: src/sympc/store/crypto_primitive_provider.py [hunk 4]
```diff
@@ -86,3 +93,31 @@ def get_state() -> str:
...
+
+        Args:
+            generate_file: when set to True generates a seperate primitive_log.json file
+        """
+        CryptoPrimitiveProvider._logging = True
+        CryptoPrimitiveProvider._generate_file = generate_file
+
+    @staticmethod
+    def stop_logging():
+        """Sets the varible to False to stop primitive logging.
```
Comment (gmuraru) [inline src/sympc/store/crypto_primitive_provider.py:109]: TYPO: ```variable```

File: src/sympc/store/crypto_primitive_provider.py [hunk 5]
```diff
@@ -13,6 +15,9 @@ class CryptoPrimitiveProvider:
     """A trusted third party should use this class to generate crypto primitives."""
 
     _func_providers: Dict[str, Callable] = {}
+    _ops_list: DefaultDict[str, List] = DefaultDict(list)
```
Comment (gmuraru) [inline src/sympc/store/crypto_primitive_provider.py:18]: I had no idea we could use the DefaultDict from typing :D.

I thought we would need to do and use that one.
```from collections import defaultdict```

and use this one

File: src/sympc/store/crypto_primitive_provider.py [hunk 6]
```diff
@@ -86,3 +93,31 @@ def get_state() -> str:
...
+        Returns:
+            json: returns the json object containing ops details.
+        """
+        CryptoPrimitiveProvider._logging = False
+        log_json = json.dumps(CryptoPrimitiveProvider._ops_list)
+        CryptoPrimitiveProvider._ops_list.clear()
+
+        if CryptoPrimitiveProvider._generate_file:
+            CryptoPrimitiveProvider._generate_file = False
+            with open("primitive_log.json", "w") as f:
```
Comment (gmuraru) [inline src/sympc/store/crypto_primitive_provider.py:120]: MAYBE: Could you have the filename in a separate variable?

File: src/sympc/store/crypto_primitive_provider.py [hunk 7]
```diff
@@ -86,3 +93,31 @@ def get_state() -> str:
         """
         res = f"Providers: {list(CryptoPrimitiveProvider._func_providers.keys())}\n"
         return res
+
+    @staticmethod
+    def start_logging(generate_file: bool = False) -> None:
```
Comment (gmuraru) [inline src/sympc/store/crypto_primitive_provider.py:98]: MAYBE: Move the ```generate_file``` arg to the ```stop_logging```. In that way, we can remove it from class.

File: src/sympc/store/crypto_primitive_provider.py (hunk 4 above)
Comment (aanurraj) [reply src/sympc/store/crypto_primitive_provider.py:109]: done

File: src/sympc/store/crypto_primitive_provider.py (hunk 7 above)
Comment (aanurraj) [reply src/sympc/store/crypto_primitive_provider.py:98]: done

**Review by gmuraru — APPROVED** [review]
LGTM!

**Pull request merged**